apps/api/v2/src/modules/slots/slots-2024-09-04/services/slots.service.ts

```diff
@@ -0,0 +1,157 @@
+          : slot.time;
+        if (!slotTimezoneAdjusted) {
+          throw new BadRequestException(
+            `Could not adjust timezone for slot ${slot.time} with timezone ${timeZone}`
+          );
+        }
+
+        return slotTimezoneAdjusted;
+      });
+    }
+
+    return slots;
+  }
+
+  adjustEndTime(endTime: string) {
+    let dateTime = DateTime.fromISO(endTime);
+    if (dateTime.hour === 0 && dateTime.minute === 0 && dateTime.second === 0) {
+      dateTime = dateTime.set({ hour: 23, minute: 59, second: 59 });
+    }
+
+    return dateTime.toISO();
+  }
+
+  async reserveSlot(input: ReserveSlotInput_2024_09_04, existingUid?: string) {
+    const eventType = await this.eventTypeRepo.getEventTypeWithHosts(input.eventTypeId);
+    if (!eventType) {
+      throw new NotFoundException(`Event Type with ID=${input.eventTypeId} not found`);
+    }
+
+    const startDate = DateTime.fromISO(input.start, { zone: "utc" });
+    if (!startDate.isValid) {
+      throw new BadRequestException("Invalid start date");
+    }
+
+    const endDate = startDate.plus({ minutes: eventType.length });
+    if (!endDate.isValid) {
+      throw new BadRequestException("Invalid end date");
+    }
+
+    const booking = await this.slotsRepo.getBookingWithAttendeesByEventTypeIdAndStart(
+      input.eventTypeId,
+      startDate.toJSDate()
+    );
+
+    if (eventType.seatsPerTimeSlot) {
+      const attendeesCount = booking?.attendees?.length;
+      if (attendeesCount) {
+        const seatsLeft = eventType.seatsPerTimeSlot - attendeesCount;
+        if (seatsLeft < 1) {
+          throw new UnprocessableEntityException(
+            `Booking with id=${input.eventTypeId} at ${input.start} has no more seats left.`
+          );
+        }
+      }
+    }
+
+    if (!eventType.seatsPerTimeSlot && booking) {
+      throw new UnprocessableEntityException(`Can't reserve a slot if the event is already booked.`);
+    }
+
+    const uid = existingUid || uuid();
+    if (eventType.userId) {
+      await this.slotsRepo.upsertSlot(
+        eventType.userId,
+        eventType.id,
+        startDate.toISO(),
+        endDate.toISO(),
+        uid,
+        eventType.seatsPerTimeSlot !== null,
+        input.reservationLength
+      );
+    } else {
+      const host = eventType.hosts[0];
+      await this.slotsRepo.upsertSlot(
+        host.userId,
+        eventType.id,
+        startDate.toISO(),
+        endDate.toISO(),
+        uid,
+        eventType.seatsPerTimeSlot !== null,
+        input.reservationLength
+      );
+    }
+
+    return uid;
+  }
+
+  async deleteSelectedSlot(uid: string) {
+    return this.slotsRepo.deleteSelectedSlots(uid);
+  }
+
+  async getEventType(input: GetSlotsInput_2024_09_04) {
+    if ("eventTypeId" in input) {
+      return this.eventTypeRepo.getEventTypeById(input.eventTypeId);
+    }
+    if ("eventTypeSlug" in input) {
+      return this.eventTypeRepo.getEventTypeBySlug(input.eventTypeSlug);
```

Comment:
Yeah we should handle this with teamId and userId respectively, or we can't trust that the function returns the intended eventType.